File: src/extras-kfilemetadata/libsimpleextractionresult.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_KFileMetaData__ExtractionResult
#define WORKAROUND_INNER_CLASS_DEFINITION_KFileMetaData__SimpleExtractionResult
#include <QList>
#include <QMultiMap>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <simpleextractionresult.h>
#include "libsimpleextractionresult.h"
#include "libsimpleextractionresult.hxx"

KFileMetaData__SimpleExtractionResult* KFileMetaData__SimpleExtractionResult_new(const libqt_string url) {
    QString url_QString = QString::fromUtf8(url.data, url.len);
    return new VirtualKFileMetaDataSimpleExtractionResult(url_QString);
}

KFileMetaData__SimpleExtractionResult* KFileMetaData__SimpleExtractionResult_new2(const KFileMetaData__SimpleExtractionResult* rhs) {
    return new VirtualKFileMetaDataSimpleExtractionResult(*rhs);
}

KFileMetaData__SimpleExtractionResult* KFileMetaData__SimpleExtractionResult_new3(const libqt_string url, const libqt_string mimetype) {
    QString url_QString = QString::fromUtf8(url.data, url.len);
    QString mimetype_QString = QString::fromUtf8(mimetype.data, mimetype.len);
    return new VirtualKFileMetaDataSimpleExtractionResult(url_QString, mimetype_QString);
}

KFileMetaData__SimpleExtractionResult* KFileMetaData__SimpleExtractionResult_new4(const libqt_string url, const libqt_string mimetype, const int* flags) {
    QString url_QString = QString::fromUtf8(url.data, url.len);
    QString mimetype_QString = QString::fromUtf8(mimetype.data, mimetype.len);
    return new VirtualKFileMetaDataSimpleExtractionResult(url_QString, mimetype_QString, (const KFileMetaData::ExtractionResult::Flags&)(*flags));
}

void KFileMetaData__SimpleExtractionResult_OperatorAssign(KFileMetaData__SimpleExtractionResult* self, const KFileMetaData__SimpleExtractionResult* rhs) {
    self->operator=(*rhs);
}

bool KFileMetaData__SimpleExtractionResult_OperatorEqual(const KFileMetaData__SimpleExtractionResult* self, const KFileMetaData__SimpleExtractionResult* rhs) {
    return (*self == *rhs);
}

void KFileMetaData__SimpleExtractionResult_Add(KFileMetaData__SimpleExtractionResult* self, int property, const QVariant* value) {
    auto* vkfilemetadata__simpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadata__simpleextractionresult && vkfilemetadata__simpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        self->add(static_cast<KFileMetaData::Property::Property>(property), *value);
    } else {
        ((VirtualKFileMetaDataSimpleExtractionResult*)self)->add(static_cast<KFileMetaData::Property::Property>(property), *value);
    }
}

void KFileMetaData__SimpleExtractionResult_AddType(KFileMetaData__SimpleExtractionResult* self, int typeVal) {
    auto* vkfilemetadata__simpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadata__simpleextractionresult && vkfilemetadata__simpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        self->addType(static_cast<KFileMetaData::Type::Type>(typeVal));
    } else {
        ((VirtualKFileMetaDataSimpleExtractionResult*)self)->addType(static_cast<KFileMetaData::Type::Type>(typeVal));
    }
}

void KFileMetaData__SimpleExtractionResult_Append(KFileMetaData__SimpleExtractionResult* self, const libqt_string text) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    auto* vkfilemetadata__simpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadata__simpleextractionresult && vkfilemetadata__simpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        self->append(text_QString);
    } else {
        ((VirtualKFileMetaDataSimpleExtractionResult*)self)->append(text_QString);
    }
}

libqt_map /* of int to libqt_list of QVariant* */ KFileMetaData__SimpleExtractionResult_Properties(const KFileMetaData__SimpleExtractionResult* self) {
    QMultiMap<KFileMetaData::Property::Property, QVariant> _ret = self->properties();
    // Convert QMultiMap<> from C++ memory to manually-managed C memory
    auto _uniqueKeys = _ret.uniqueKeys();
    auto _numUniqueKeys = _uniqueKeys.size();
    int* _karr = static_cast<int*>(malloc(sizeof(int) * _numUniqueKeys));
    libqt_list* _varr = static_cast<libqt_list*>(malloc(sizeof(libqt_list) * _numUniqueKeys));
    for (auto i = 0; i < _numUniqueKeys; ++i) {
        KFileMetaData::Property::Property key = _uniqueKeys[i];
        _karr[i] = static_cast<int>(key);
        QList<QVariant> values = _ret.values(key);
        size_t numValues = values.size();
        QVariant** _array = static_cast<QVariant**>(malloc(sizeof(QVariant*) * numValues));
        for (size_t j = 0; j < numValues; ++j) {
            _array[j] = new QVariant(values[j]);
        }
        _varr[i].len = numValues;
        _varr[i].data = static_cast<void*>(_array);
    }
    libqt_map _out;
    _out.len = _numUniqueKeys;
    _out.keys = static_cast<void*>(_karr);
    _out.values = static_cast<void*>(_varr);
    return _out;
}

libqt_string KFileMetaData__SimpleExtractionResult_Text(const KFileMetaData__SimpleExtractionResult* self) {
    QString _ret = self->text();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of int */ KFileMetaData__SimpleExtractionResult_Types(const KFileMetaData__SimpleExtractionResult* self) {
    QList<KFileMetaData::Type::Type> _ret = self->types();
    // Convert QList<> from C++ memory to manually-managed C memory
    int* _arr = static_cast<int*>(malloc(sizeof(int) * (_ret.size())));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = static_cast<int>(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

// Base class handler implementation
void KFileMetaData__SimpleExtractionResult_QBaseAdd(KFileMetaData__SimpleExtractionResult* self, int property, const QVariant* value) {
    auto* vkfilemetadatasimpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadatasimpleextractionresult && vkfilemetadatasimpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        vkfilemetadatasimpleextractionresult->setKFileMetaData__SimpleExtractionResult_Add_IsBase(true);
        vkfilemetadatasimpleextractionresult->add(static_cast<KFileMetaData::Property::Property>(property), *value);
    } else {
        self->KFileMetaData::SimpleExtractionResult::add(static_cast<KFileMetaData::Property::Property>(property), *value);
    }
}

// Auxiliary method to allow providing re-implementation
void KFileMetaData__SimpleExtractionResult_OnAdd(KFileMetaData__SimpleExtractionResult* self, intptr_t slot) {
    auto* vkfilemetadatasimpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadatasimpleextractionresult && vkfilemetadatasimpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        vkfilemetadatasimpleextractionresult->setKFileMetaData__SimpleExtractionResult_Add_Callback(reinterpret_cast<VirtualKFileMetaDataSimpleExtractionResult::KFileMetaData__SimpleExtractionResult_Add_Callback>(slot));
    }
}

// Base class handler implementation
void KFileMetaData__SimpleExtractionResult_QBaseAddType(KFileMetaData__SimpleExtractionResult* self, int typeVal) {
    auto* vkfilemetadatasimpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadatasimpleextractionresult && vkfilemetadatasimpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        vkfilemetadatasimpleextractionresult->setKFileMetaData__SimpleExtractionResult_AddType_IsBase(true);
        vkfilemetadatasimpleextractionresult->addType(static_cast<KFileMetaData::Type::Type>(typeVal));
    } else {
        self->KFileMetaData::SimpleExtractionResult::addType(static_cast<KFileMetaData::Type::Type>(typeVal));
    }
}

// Auxiliary method to allow providing re-implementation
void KFileMetaData__SimpleExtractionResult_OnAddType(KFileMetaData__SimpleExtractionResult* self, intptr_t slot) {
    auto* vkfilemetadatasimpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadatasimpleextractionresult && vkfilemetadatasimpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        vkfilemetadatasimpleextractionresult->setKFileMetaData__SimpleExtractionResult_AddType_Callback(reinterpret_cast<VirtualKFileMetaDataSimpleExtractionResult::KFileMetaData__SimpleExtractionResult_AddType_Callback>(slot));
    }
}

// Base class handler implementation
void KFileMetaData__SimpleExtractionResult_QBaseAppend(KFileMetaData__SimpleExtractionResult* self, const libqt_string text) {
    auto* vkfilemetadatasimpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    QString text_QString = QString::fromUtf8(text.data, text.len);
    if (vkfilemetadatasimpleextractionresult && vkfilemetadatasimpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        vkfilemetadatasimpleextractionresult->setKFileMetaData__SimpleExtractionResult_Append_IsBase(true);
        vkfilemetadatasimpleextractionresult->append(text_QString);
    } else {
        self->KFileMetaData::SimpleExtractionResult::append(text_QString);
    }
}

// Auxiliary method to allow providing re-implementation
void KFileMetaData__SimpleExtractionResult_OnAppend(KFileMetaData__SimpleExtractionResult* self, intptr_t slot) {
    auto* vkfilemetadatasimpleextractionresult = dynamic_cast<VirtualKFileMetaDataSimpleExtractionResult*>(self);
    if (vkfilemetadatasimpleextractionresult && vkfilemetadatasimpleextractionresult->isVirtualKFileMetaDataSimpleExtractionResult) {
        vkfilemetadatasimpleextractionresult->setKFileMetaData__SimpleExtractionResult_Append_Callback(reinterpret_cast<VirtualKFileMetaDataSimpleExtractionResult::KFileMetaData__SimpleExtractionResult_Append_Callback>(slot));
    }
}

void KFileMetaData__SimpleExtractionResult_Delete(KFileMetaData__SimpleExtractionResult* self) {
    delete self;
}

File: include/extras-kfilemetadata/libsimpleextractionresult.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QVariant = @import("libqt6").QVariant;
const embeddedimagedata_enums = @import("libembeddedimagedata.zig").enums;
const extractionresult_enums = @import("libextractionresult.zig").enums;
const properties_enums = @import("libproperties.zig").enums;
const std = @import("std");
const types_enums = @import("libtypes.zig").enums;
const ArrayMap_i32_SliceQVariant = std.array_hash_map.Auto(i32, []QVariant);
const ArrayMap_i32_u8 = std.array_hash_map.Auto(i32, []u8);

/// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html)
pub const KFileMetaData__SimpleExtractionResult = extern struct {
    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.KFileMetaData__SimpleExtractionResult,

    pub const _is_KFileMetaData__SimpleExtractionResult = {};
    pub const _is_KFileMetaData__ExtractionResult = {};

    /// ### DEPRECATED: Use `new` instead
    ///
    pub const New = new;

    /// Allocate a new KFileMetaData::SimpleExtractionResult object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` url: []const u8 `
    ///
    pub fn new(url: []const u8) KFileMetaData__SimpleExtractionResult {
        const url_str = qtc.libqt_string{
            .len = url.len,
            .data = url.ptr,
        };
        return .{ .ptr = qtc.KFileMetaData__SimpleExtractionResult_new(url_str) };
    }

    /// ### DEPRECATED: Use `new2` instead
    ///
    pub const New2 = new2;

    /// Allocate a new KFileMetaData::SimpleExtractionResult object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` rhs: KFileMetaData__SimpleExtractionResult `
    ///
    pub fn new2(rhs: anytype) KFileMetaData__SimpleExtractionResult {
        comptime _ = @TypeOf(rhs)._is_KFileMetaData__SimpleExtractionResult;
        return .{ .ptr = qtc.KFileMetaData__SimpleExtractionResult_new2(@ptrCast(rhs.ptr)) };
    }

    /// ### DEPRECATED: Use `new3` instead
    ///
    pub const New3 = new3;

    /// Allocate a new KFileMetaData::SimpleExtractionResult object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` url: []const u8 `
    ///
    /// ` mimetype: []const u8 `
    ///
    pub fn new3(url: []const u8, mimetype: []const u8) KFileMetaData__SimpleExtractionResult {
        const url_str = qtc.libqt_string{
            .len = url.len,
            .data = url.ptr,
        };
        const mimetype_str = qtc.libqt_string{
            .len = mimetype.len,
            .data = mimetype.ptr,
        };
        return .{ .ptr = qtc.KFileMetaData__SimpleExtractionResult_new3(url_str, mimetype_str) };
    }

    /// ### DEPRECATED: Use `new4` instead
    ///
    pub const New4 = new4;

    /// Allocate a new KFileMetaData::SimpleExtractionResult object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` url: []const u8 `
    ///
    /// ` mimetype: []const u8 `
    ///
    /// ` flags: *const flag of extractionresult_enums.Flag `
    ///
    pub fn new4(url: []const u8, mimetype: []const u8, flags: *const i32) KFileMetaData__SimpleExtractionResult {
        const url_str = qtc.libqt_string{
            .len = url.len,
            .data = url.ptr,
        };
        const mimetype_str = qtc.libqt_string{
            .len = mimetype.len,
            .data = mimetype.ptr,
        };
        return .{ .ptr = qtc.KFileMetaData__SimpleExtractionResult_new4(url_str, mimetype_str, @ptrCast(flags)) };
    }

    /// ### DEPRECATED: Use `operatorAssign` instead
    ///
    pub const OperatorAssign = operatorAssign;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#operator-eq)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` rhs: KFileMetaData__SimpleExtractionResult `
    ///
    pub fn operatorAssign(self: KFileMetaData__SimpleExtractionResult, rhs: anytype) void {
        comptime _ = @TypeOf(rhs)._is_KFileMetaData__SimpleExtractionResult;
        qtc.KFileMetaData__SimpleExtractionResult_OperatorAssign(@ptrCast(self.ptr), @ptrCast(rhs.ptr));
    }

    /// ### DEPRECATED: Use `operatorEqual` instead
    ///
    pub const OperatorEqual = operatorEqual;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#operator-eq-eq)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` rhs: KFileMetaData__SimpleExtractionResult `
    ///
    pub fn operatorEqual(self: KFileMetaData__SimpleExtractionResult, rhs: anytype) bool {
        comptime _ = @TypeOf(rhs)._is_KFileMetaData__SimpleExtractionResult;
        return qtc.KFileMetaData__SimpleExtractionResult_OperatorEqual(@ptrCast(self.ptr), @ptrCast(rhs.ptr));
    }

    /// ### DEPRECATED: Use `add` instead
    ///
    pub const Add = add;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#add)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` property: properties_enums.Property `
    ///
    /// ` value: QVariant `
    ///
    pub fn add(self: KFileMetaData__SimpleExtractionResult, property: i32, value: anytype) void {
        comptime _ = @TypeOf(value)._is_QVariant;
        qtc.KFileMetaData__SimpleExtractionResult_Add(@ptrCast(self.ptr), @bitCast(property), @ptrCast(value.ptr));
    }

    /// ### DEPRECATED: Use `onAdd` instead
    ///
    pub const OnAdd = onAdd;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#add)
    ///
    /// Allows for overriding the related default method
    ///
    /// ## Parameters:
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` callback: *const fn (self: KFileMetaData__SimpleExtractionResult, property: properties_enums.Property, value: QVariant) callconv(.c) void `
    ///
    pub fn onAdd(self: KFileMetaData__SimpleExtractionResult, callback: *const fn (KFileMetaData__SimpleExtractionResult, i32, QVariant) callconv(.c) void) void {
        qtc.KFileMetaData__SimpleExtractionResult_OnAdd(@ptrCast(self.ptr), @bitCast(@intFromPtr(callback)));
    }

    /// ### DEPRECATED: Use `superAdd` instead
    ///
    pub const SuperAdd = superAdd;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#add)
    ///
    /// Base class method implementation
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` property: properties_enums.Property `
    ///
    /// ` value: QVariant `
    ///
    pub fn superAdd(self: KFileMetaData__SimpleExtractionResult, property: i32, value: anytype) void {
        comptime _ = @TypeOf(value)._is_QVariant;
        qtc.KFileMetaData__SimpleExtractionResult_SuperAdd(@ptrCast(self.ptr), @bitCast(property), @ptrCast(value.ptr));
    }

    /// ### DEPRECATED: Use `addType` instead
    ///
    pub const AddType = addType;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#addType)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` typeVal: types_enums.Type `
    ///
    pub fn addType(self: KFileMetaData__SimpleExtractionResult, typeVal: i32) void {
        qtc.KFileMetaData__SimpleExtractionResult_AddType(@ptrCast(self.ptr), @bitCast(typeVal));
    }

    /// ### DEPRECATED: Use `onAddType` instead
    ///
    pub const OnAddType = onAddType;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#addType)
    ///
    /// Allows for overriding the related default method
    ///
    /// ## Parameters:
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` callback: *const fn (self: KFileMetaData__SimpleExtractionResult, typeVal: types_enums.Type) callconv(.c) void `
    ///
    pub fn onAddType(self: KFileMetaData__SimpleExtractionResult, callback: *const fn (KFileMetaData__SimpleExtractionResult, i32) callconv(.c) void) void {
        qtc.KFileMetaData__SimpleExtractionResult_OnAddType(@ptrCast(self.ptr), @bitCast(@intFromPtr(callback)));
    }

    /// ### DEPRECATED: Use `superAddType` instead
    ///
    pub const SuperAddType = superAddType;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#addType)
    ///
    /// Base class method implementation
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` typeVal: types_enums.Type `
    ///
    pub fn superAddType(self: KFileMetaData__SimpleExtractionResult, typeVal: i32) void {
        qtc.KFileMetaData__SimpleExtractionResult_SuperAddType(@ptrCast(self.ptr), @bitCast(typeVal));
    }

    /// ### DEPRECATED: Use `append` instead
    ///
    pub const Append = append;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#append)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` _text: []const u8 `
    ///
    pub fn append(self: KFileMetaData__SimpleExtractionResult, _text: []const u8) void {
        const text_str = qtc.libqt_string{
            .len = _text.len,
            .data = _text.ptr,
        };
        qtc.KFileMetaData__SimpleExtractionResult_Append(@ptrCast(self.ptr), text_str);
    }

    /// ### DEPRECATED: Use `onAppend` instead
    ///
    pub const OnAppend = onAppend;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#append)
    ///
    /// Allows for overriding the related default method
    ///
    /// ## Parameters:
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` callback: *const fn (self: KFileMetaData__SimpleExtractionResult, text: [*:0]const u8) callconv(.c) void `
    ///
    pub fn onAppend(self: KFileMetaData__SimpleExtractionResult, callback: *const fn (KFileMetaData__SimpleExtractionResult, [*:0]const u8) callconv(.c) void) void {
        qtc.KFileMetaData__SimpleExtractionResult_OnAppend(@ptrCast(self.ptr), @bitCast(@intFromPtr(callback)));
    }

    /// ### DEPRECATED: Use `superAppend` instead
    ///
    pub const SuperAppend = superAppend;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#append)
    ///
    /// Base class method implementation
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` _text: []const u8 `
    ///
    pub fn superAppend(self: KFileMetaData__SimpleExtractionResult, _text: []const u8) void {
        const text_str = qtc.libqt_string{
            .len = _text.len,
            .data = _text.ptr,
        };
        qtc.KFileMetaData__SimpleExtractionResult_SuperAppend(@ptrCast(self.ptr), text_str);
    }

    /// ### DEPRECATED: Use `properties` instead
    ///
    pub const Properties = properties;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#properties)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ## Returns:
    ///
    /// ` ArrayMap_i32_SliceQVariant (key: properties_enums.Property) `
    ///
    pub fn properties(self: KFileMetaData__SimpleExtractionResult, allocator: std.mem.Allocator) ArrayMap_i32_SliceQVariant {
        const _map: qtc.libqt_map = qtc.KFileMetaData__SimpleExtractionResult_Properties(@ptrCast(self.ptr));
        var _ret: ArrayMap_i32_SliceQVariant = .empty;
        _ret.ensureTotalCapacity(allocator, @intCast(_map.len)) catch @panic("KFileMetaData__SimpleExtractionResult.properties: Total capacity allocation failed");
        defer {
            const _values: [*]qtc.libqt_list = @ptrCast(@alignCast(_map.values));
            for (0.._map.len) |i| {
                qtc.libqt_free(_values[i].data);
            }
            qtc.libqt_free(_map.keys);
            qtc.libqt_free(_map.values);
        }
        const _keys: [*]i32 = @ptrCast(@alignCast(_map.keys));
        const _values: [*]qtc.libqt_list = @ptrCast(@alignCast(_map.values));
        var i: usize = 0;
        while (i < _map.len) : (i += 1) {
            const _key = _keys[i];
            const _value = _values[i];
            const _value_slice = allocator.alloc(QVariant, _value.len) catch @panic("KFileMetaData__SimpleExtractionResult.properties: Memory allocation failed");
            const _value_data: [*]QtC.QVariant = @ptrCast(@alignCast(_value.data));
            for (0.._value.len) |j|
                _value_slice[j] = .{ .ptr = _value_data[j] };
            _ret.putAssumeCapacity(_key, _value_slice);
        }
        return _ret;
    }

    /// ### DEPRECATED: Use `text` instead
    ///
    pub const Text = text;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#text)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn text(self: KFileMetaData__SimpleExtractionResult, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.KFileMetaData__SimpleExtractionResult_Text(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KFileMetaData__SimpleExtractionResult.text: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `types` instead
    ///
    pub const Types = types;

    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-simpleextractionresult.html#types)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ## Returns:
    ///
    /// ` []types_enums.Type `
    ///
    pub fn types(self: KFileMetaData__SimpleExtractionResult, allocator: std.mem.Allocator) []i32 {
        const _arr: qtc.libqt_list = qtc.KFileMetaData__SimpleExtractionResult_Types(@ptrCast(self.ptr));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(i32, _arr.len) catch @panic("KFileMetaData__SimpleExtractionResult.types: Memory allocation failed");
        const _data_val: [*]i32 = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data_val[0.._arr.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `inputUrl` instead
    ///
    pub const InputUrl = inputUrl;

    /// Inherited from KFileMetaData::ExtractionResult
    ///
    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-extractionresult.html#inputUrl)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn inputUrl(self: KFileMetaData__SimpleExtractionResult, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.KFileMetaData__ExtractionResult_InputUrl(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KFileMetaData__SimpleExtractionResult.inputUrl: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `inputMimetype` instead
    ///
    pub const InputMimetype = inputMimetype;

    /// Inherited from KFileMetaData::ExtractionResult
    ///
    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-extractionresult.html#inputMimetype)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn inputMimetype(self: KFileMetaData__SimpleExtractionResult, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.KFileMetaData__ExtractionResult_InputMimetype(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KFileMetaData__SimpleExtractionResult.inputMimetype: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `inputFlags` instead
    ///
    pub const InputFlags = inputFlags;

    /// Inherited from KFileMetaData::ExtractionResult
    ///
    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-extractionresult.html#inputFlags)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ## Returns:
    ///
    /// ` flag of extractionresult_enums.Flag `
    ///
    pub fn inputFlags(self: KFileMetaData__SimpleExtractionResult) i32 {
        return qtc.KFileMetaData__ExtractionResult_InputFlags(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `imageData` instead
    ///
    pub const ImageData = imageData;

    /// Inherited from KFileMetaData::ExtractionResult
    ///
    /// ### [Upstream resources](https://api.kde.org/kfilemetadata-extractionresult.html#imageData)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ## Returns:
    ///
    /// ` ArrayMap_i32_u8 (key: embeddedimagedata_enums.ImageType) `
    ///
    pub fn imageData(self: KFileMetaData__SimpleExtractionResult, allocator: std.mem.Allocator) ArrayMap_i32_u8 {
        const _map: qtc.libqt_map = qtc.KFileMetaData__ExtractionResult_ImageData(@ptrCast(self.ptr));
        var _ret: ArrayMap_i32_u8 = .empty;
        _ret.ensureTotalCapacity(allocator, @intCast(_map.len)) catch @panic("KFileMetaData__SimpleExtractionResult.imageData: Total capacity allocation failed");
        defer {
            const _values: [*]qtc.libqt_string = @ptrCast(@alignCast(_map.values));
            for (0.._map.len) |i| {
                qtc.libqt_free(_values[i].data);
            }
            qtc.libqt_free(_map.keys);
            qtc.libqt_free(_map.values);
        }
        const _keys: [*]i32 = @ptrCast(@alignCast(_map.keys));
        const _values: [*]qtc.libqt_string = @ptrCast(@alignCast(_map.values));
        var i: usize = 0;
        while (i < _map.len) : (i += 1) {
            const _key = _keys[i];
            const _value = _values[i];
            const _value_slice = allocator.alloc(u8, _value.len) catch @panic("KFileMetaData__SimpleExtractionResult.imageData: Memory allocation failed");
            @memcpy(_value_slice, _value.data);
            _ret.putAssumeCapacity(_key, _value_slice);
        }
        return _ret;
    }

    /// ### DEPRECATED: Use `delete` instead
    ///
    pub const Delete = delete;

    /// Delete this object from C++ memory
    ///
    /// ## Parameter:
    ///
    /// ` self: KFileMetaData__SimpleExtractionResult `
    ///
    pub fn delete(self: KFileMetaData__SimpleExtractionResult) void {
        qtc.KFileMetaData__SimpleExtractionResult_Delete(@ptrCast(self.ptr));
    }
};
